### PR TITLE
Updated CPAN type spec

### PR DESCRIPTION
**Pull request by giterlizzi**
This PR is for the new version of the purl-type CPAN specification and includes new updated tests.

https://github.com/giterlizzi/perl-URI-PackageURL/issues/8

In brief, this PR confirms the rules for generating a PURL string for CPAN.

Previously it was possible to create PURL strings in two ways:
- author (namespace) and distribution (name)
- module (name component).

Using only the module could create ambiguity, because in CPAN a distribution can have several authors (and co-authors) over time and does not allow precise identification of the artifact.

It was clarified that the author of the "distribution" is mandatory, allowing an artifact and its URL to be identified along with its version.

```console
$ purl-tool pkg:cpan/GDT/URI-PackageURL@2.22 --download-url
https://www.cpan.org/authors/id/G/GD/GDT/URI-PackageURL-2.22.tar.gz
```

### CODE REVIEW

File: test-suite-data.json [hunk 1]
```diff
@@ -576,31 +576,31 @@
...
+    "version": "6.76",
     "qualifiers": null,
     "subpath": null,
     "is_invalid": false
   },
   {
-    "description": "cpan module name are case sensitive",
-    "purl": "pkg:cpan/URI::PackageURL@2.11",
-    "canonical_purl": "pkg:cpan/URI::PackageURL@2.11",
+    "description": "CPAN distribution name are case sensitive",
```
Comment (sjn) [inline test-suite-data.json:591]: `names` in plural, or `is` in singular

File: test-suite-data.json [hunk 2]
```diff
@@ -612,38 +612,14 @@
     "is_invalid": true
   },
   {
-    "description": "cpan distribution name like module name",
-    "purl": "pkg:cpan/GDT/URI::PackageURL@2.11",
-    "canonical_purl": "pkg:cpan/GDT/URI::PackageURL",
+    "description": "CPAN distribution in BackPan repository",
```
Comment (sjn) [inline test-suite-data.json:615]: BackPAN

**Review by sjn — COMMENTED** [review]
Looks good! Just a few tweaks, and one question.

Would it make sense to add some failing tests?

e.g. the previous format with `Module::Names`, and perhaps a couple more uses, with a short description on why it is invalid?

File: test-suite-data.json (hunk 2 above)
Comment (giterlizzi) [reply test-suite-data.json]: FIXED typo

File: test-suite-data.json (hunk 1 above)
Comment (giterlizzi) [reply test-suite-data.json]: FIXED typo

File: test-suite-data.json [hunk 3]
```diff
@@ -623,6 +623,18 @@
...
+    "purl": "pkg:cpan/GDT/URI::PackageURL",
+    "canonical_purl": "pkg:cpan/GDT/URI::PackageURL",
+    "type": "cpan",
+    "namespace": "GDT",
+    "name": "URI::PackageURL",
+    "version": null,
+    "qualifiers": null,
+    "subpath": null,
+    "is_invalid": true
+  },
```
Comment (sjn) [inline test-suite-data.json:637]: Would it be sensible to also add a failing test for the previously accepted pURLs? e.g.
```
  {
    "description": "invalid CPAN with just the module name",
    "purl": "pkg:cpan/URI::PackageURL",
    "canonical_purl": "pkg:cpan/URI::PackageURL",
    "type": "cpan",
    "namespace": null,
    "name": "URI::PackageURL",
    "version": null,
    "qualifiers": null,
    "subpath": null,
    "is_invalid": true
  },
```
 …and maybe the variants that include a version number?

Comment (giterlizzi) [reply test-suite-data.json:637]: right, the test "invalid CPAN distribution without author" cover only the "distribution" naming. added ;)

**Comment by pombredanne** [comment]
@giterlizzi Thanks. We need to make sure we are all clear first wrt. the encoding in the core spec and that is getting close to merge ready in #398

**Review by giterlizzi — COMMENTED** [review]
the PR now includes these new tests for the new CPAN specification:
- valid CPAN purl
- CPAN distribution in BackPAN repository
- invalid CPAN distribution name as module name

File: test-suite-data.json [hunk 4]
```diff
@@ -624,28 +636,28 @@
...
     "type": "cpan",
     "namespace": null,
-    "name": "URI",
+    "name": "DBD::SQLite",
     "version": null,
     "qualifiers": null,
     "subpath": null,
-    "is_invalid": false
+    "is_invalid": true
   },
```
Comment (sjn) [inline test-suite-data.json]: Would it be sensible to add a few more tests for corner cases?

* Distribution name with lowercase characters
* Distribution name which contains `::`

Comment (sjn) [reply test-suite-data.json]: Looks otherwise good! :-)

File: PURL-TYPES.rst [hunk 5]
```diff
@@ -170,15 +170,12 @@ conda
...
 
 - The default repository is ``https://www.cpan.org/``.
-- The ``namespace``:
-  - To refer to a CPAN distribution name, the ``namespace`` MUST be present. In this case, the namespace is the CPAN id of the author/publisher. It MUST be written uppercase, followed by the distribution name in the ``name`` component. A distribution name MUST NOT contain the string ``::``.
-  - To refer to a CPAN module, the ``namespace`` MUST be absent. The module name MAY contain zero or more ``::`` strings, and the module name MUST NOT contain a ``-``
-
-- The ``name`` is the module or distribution name and is case sensitive.
-- The ``version`` is the module or distribution version.
+- The ``namespace`` is the CPAN id of the author/publisher. It MUST be written uppercase and is required.
+- The ``name`` is the distribution name and is case sensitive.
```
Comment (sjn) [inline PURL-TYPES.rst]: Not sure, but would it make sense to still point out that…

> A distribution name MUST NOT contain the string ``::``.

…as earlier? I get it that it's strictly speaking not necessary, but as a hint for new users to understand the difference between a module name and a distribution name, I think it can be helpful.

(I have no strong opinions on this idea, and leave it to you to decide if this goes in or not)

**Review by sjn — APPROVED** [review]
Looks good to me, with a few final optional suggestions.

File: PURL-TYPES.rst (hunk 5 above)
Comment (giterlizzi) [reply PURL-TYPES.rst]: I agree, adding this sentence (`A distribution name MUST NOT contain the string ::`) reinforces the concept that only the distribution MUST be used.

Comment (giterlizzi) [reply PURL-TYPES.rst]: "name" component description updated

**Review by pombredanne — CHANGES_REQUESTED** [review]
Looking good, and this is a big change... could it make sense to also add a FAQ entry and give hints to implementers what they can do when parsing older `::` PURLs?

**Comment by pombredanne** [comment]
After the merge of PR #514, PURL types are now defined in JSON: :innocent: :grin: 
- See  #514

With the new approach... this PR needs to be updated. Could you look into this? Thanks for your understanding!

**Comment by giterlizzi** [comment]
@sjn @pombredanne Hi, i have updated the CPAN type spec and tests.

**Comment by giterlizzi** [comment]
> Looking good, and this is a big change... could it make sense to also add a FAQ entry and give hints to implementers what they can do when parsing older `::` PURLs?

If @sjn agrees, I will add the "notes" field with this sentence:

The previous CPAN PURL type specification allowed module names (e.g. URI::PackageURL) to be used as PURL "names" while also omitting the PURL "namespace". The parser MUST emit an error when a module is specified as a PURL "name" or detect "::" characters.

**Review by pombredanne — APPROVED** [review]
@giterlizzi this looks all good to me... is there more that you need to add to this?

**Comment by giterlizzi** [comment]
Nothing to add. Thanks ;)

**Comment by sjn** [comment]
LGTM!
